### PR TITLE
Added Error to return type of async validator

### PR DESCRIPTION
**Pull request by vjau**
This PR changes the Typescript definition.
When using an async validator, errors are either throwed or returned.
The initial typing only considered the case when Error where throwed. This PR add a union type including Error on the returned Promise.
The test is also modified by returning an Error

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by vjau** [comment]
Sorry, i have seen that there is already a pending PR addressing what this PR is doing. #58 
However, #58 is more ambitious (i have actually tried to do what #58 try to do the whole day without success), i think this more limited fix could be merged for now. 
Thank you.

File: index.test-d.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@ app.register(fastifyBasicAuth, {
     expectType<FastifyRequest>(req)
     expectType<FastifyReply>(reply)
     expectType<FastifyInstance>(this)
+    return new Error("unauthorized")
```
Comment (climba03003) [inline index.test-d.ts:22]: Can you split it in separate test? So, it test both `void` and `Error`.

**Review by climba03003 — CHANGES_REQUESTED** [review]

**Comment by vjau** [comment]
I have splitted the test as requested

**Review by mcollina — APPROVED** [review]
lgtm

**Review by climba03003 — APPROVED** [review]
LGTM.

**Pull request merged**